crates/shared/src/code_simulation.rs

```diff
@@ -0,0 +1,190 @@
+//! Abstraction for simulating calls with overrides.
+
+use crate::{
+    tenderly_api::{SimulationKind, SimulationRequest, TenderlyApi},
+    transport::extensions::{EthExt as _, StateOverrides},
+    Web3,
+};
+use anyhow::{bail, Context as _, Result};
+use std::sync::Arc;
+use web3::types::{BlockNumber, CallRequest};
+
+/// Simulate a call with state overrides.
+#[async_trait::async_trait]
+pub trait CodeSimulating: Send + Sync + 'static {
+    async fn simulate(&self, call: CallRequest, overrides: StateOverrides) -> Result<Vec<u8>>;
+}
+
+#[async_trait::async_trait]
+impl CodeSimulating for Web3 {
+    async fn simulate(&self, call: CallRequest, overrides: StateOverrides) -> Result<Vec<u8>> {
+        Ok(self
+            .eth()
+            .call_with_state_overrides(call, BlockNumber::Latest.into(), overrides)
+            .await?
+            .0)
+    }
+}
+
+pub struct TenderlyCodeSimlator {
+    tenderly: Arc<dyn TenderlyApi>,
+    network_id: String,
+    save: bool,
```

Comment:
nit: To be more consistent you could call it `save_if_succeeds`.
Alternatively you could merge the arguments into one:
```rust
enum Saving { Always, OnSuccess, OnFailure }
```

Reply:
I just kept it to have the exact same names as the fields in the Tenderly API. Will make it more consistent.